text/0034-ternary-hash.md

```diff
@@ -0,0 +1,415 @@
+it. This RFC relies on the the presence of the types `TritBuf` and `Trits`, which are assumed to be owning and borrowing
+collections of binary-coded ternary in the `T1B1` encoding (one trit per byte), similar to
+[PathBuf](https://doc.rust-lang.org/std/path/struct.PathBuf.html) and
+[Path](https://doc.rust-lang.org/std/path/struct.Path.html).
+
+```rust
+/// The common interface of cryptographic hash functions that follow the sponge construction and that act on ternary.
+trait Sponge {
+    /// The expected length of the input to the sponge.
+    const IN_LEN: usize;
+
+    /// The length of the hash squeezed from the sponge.
+    const OUT_LEN: usize;
+
+    /// An error indicating a that a failure has occured during `absorb`.
+    type Error;
+
+    /// Absorb `input` into the sponge.
+    fn absorb(&mut self, input: &Trits) -> Result<(), Self::Error>;
+
+    /// Reset the inner state of the sponge.
+    fn reset(&mut self);
+
+    /// Squeeze the sponge into a buffer
+    fn squeeze_into(&mut self, buf: &mut Trits) -> Result<(), Self::Error>;
+
+    /// Convenience function using `Sponge::squeeze_into` to to return an owned version of the hash.
+    fn squeeze(&mut self) -> Result<TritBuf, Self::Error> {
+        let mut output = TritBuf::zeros(Self::OUT_LEN);
+        self.squeeze_into(&mut output)?;
+        Ok(output)
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge into a buffer, and reset the sponge in one go.
+    fn digest_into(&mut self, input: &Trits, buf: &mut Trits) -> Result<(), Self::Error> {
+        self.absorb(input)?;
+        self.squeeze_into(buf)?;
+        self.reset();
+        Ok(())
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge, and reset the sponge in one go.
+    /// Returns an owned versin of the hash.
+    fn digest(&mut self, input: &Trits) -> Result<TritBuf, Self::Error> {
+        self.absorb(input)?;
+        let output = self.squeeze()?;
+        self.reset();
+        Ok(output)
+    }
+}
+```
+
+Following the sponge metaphor, an input reference provided by the user is `absorb`ed, and an output will be `squeeze`d
+from the data structure. `digest` is a convenience method calling `absorb` and `squeeze` in one go. The `*_into`
+versions of these methods are for passing a buffer into which the calculated hashes are written.  The internal state
+will not be cleared unless `reset` is called.
+
+### Design of `CurlP`
+
+`CurlP` is designed as a hash function that acts on a `T1B1` binary-encoded ternary buffer, with a hash length of `243`
+trits and an inner state of `729` trits, which we take as constants:
+
+```rust
+const HASH_LENGTH: usize = 243;
+const STATE_LENGTH: usize = HASH_LENGTH * 3;
+```
+
+In addition, a lookup table is used as part of the absorption step.
+2 is used to pad the table since there are only 9 combinations should be taken:
```

Comment:
the explanation is probably a little bit confusing to the reader. The reason for the pad is that the index producing function is undefined for {3, 7}

Reply:
Clarified it.